File: gradle/build-logic/src/test/kotlin/dev/erst/fingrind/buildlogic/ReviewedSurfacePolicyContractTest.kt
package dev.erst.fingrind.buildlogic

import java.nio.file.Path
import java.time.LocalDate
import kotlin.test.Test
import kotlin.test.assertEquals
import tools.jackson.databind.JsonNode
import tools.jackson.databind.json.JsonMapper

@Suppress("DEPRECATION")
class ReviewedSurfacePolicyContractTest {
    private val objectMapper = JsonMapper.builder().build()
    private val repositoryRoot = Path.of("").toAbsolutePath().normalize().parent.parent
    private val contractDirectory =
        repositoryRoot.resolve("scripts/structural_governance/reviewed_surface_policy_contract")

    @Test
    fun reviewedSurfacePolicyContract_casesStayAlignedWithKotlinWaiverSemantics() {
        val profiles = profilesByDocumentType()
        contractCases().forEach { case ->
            when (case.path("caseType").requiredText()) {
                "definition" -> assertDefinitionCase(case, profiles)
                "runtime" -> assertRuntimeCase(case, profiles)
                "orphan" -> assertOrphanCase(case, profiles)
                else ->
                    error(
                        "Unsupported reviewed-surface policy contract case type ${case.path("caseType").requiredText()}."
                    )
            }
        }
    }

    private fun contractDocuments(): List<JsonNode> =
        contractDirectory
            .toFile()
            .listFiles { file -> file.isFile && file.extension == "json" }
            ?.sortedBy { it.name }
            ?.map(objectMapper::readTree)
            ?: error("No reviewed-surface policy contract cases found in $contractDirectory")

    private fun contractCases(): List<JsonNode> =
        contractDocuments().filter { it.path("documentType").requiredText() == "case" }

    private fun profilesByDocumentType(): Map<String, Map<String, JsonNode>> =
        contractDocuments()
            .filter { it.path("documentType").requiredText() != "case" }
            .associate { document ->
                document.path("documentType").requiredText() to
                    document
                        .path("profiles")
                        .properties()
                        .asSequence()
                        .associate { entry -> entry.key to entry.value }
            }

    private fun profile(
        profiles: Map<String, Map<String, JsonNode>>,
        documentType: String,
        profileName: String,
    ): JsonNode =
        profiles[documentType]?.get(profileName)
            ?: error("Missing reviewed-surface policy contract profile $documentType/$profileName")

    private fun assertDefinitionCase(
        case: JsonNode,
        profiles: Map<String, Map<String, JsonNode>>,
    ) {
        val actual =
            reviewedSurfaceDefinitionViolations(
                reviewedSurface = reviewedSurface(case, profiles),
                defaultBudget = defaultBudget(case, profiles),
            ).map(::normalizeViolation)
        assertEquals(
            expectedDescriptors(case),
            actual,
            "Definition case ${case.path("id").requiredText()} drifted from the shared reviewed-surface policy contract.",
        )
    }

    private fun assertRuntimeCase(
        case: JsonNode,
        profiles: Map<String, Map<String, JsonNode>>,
    ) {
        val actual =
            reviewedSurfaceViolations(
                relativePath = case.path("relativePath").requiredText(),
                metrics = liveMetrics(case, profiles),
                reviewedSurface = reviewedSurface(case, profiles),
                defaultBudget = defaultBudget(case, profiles),
                currentDate = LocalDate.parse(case.path("currentDate").requiredText()),
            ).map(::normalizeViolation)
        assertEquals(
            expectedDescriptors(case),
            actual,
            "Runtime case ${case.path("id").requiredText()} drifted from the shared reviewed-surface policy contract.",
        )
    }

    private fun assertOrphanCase(
        case: JsonNode,
        profiles: Map<String, Map<String, JsonNode>>,
    ) {
        val reviewedSurface = reviewedSurface(case, profiles)
        val actual =
            missingReviewedSurfaceViolations(
                reviewedSurfaces = listOf(reviewedSurface),
                projectScope = "shared-policy-contract",
                existingRelativePaths = textArray(case.path("existingRelativePaths")).toSet(),
            ).map(::normalizeViolation)
        assertEquals(
            expectedDescriptors(case),
            actual,
            "Orphan case ${case.path("id").requiredText()} drifted from the shared reviewed-surface policy contract.",
        )
    }

    private fun expectedDescriptors(case: JsonNode): List<String> = textArray(case.path("expectedDescriptors"))

    private fun textArray(node: JsonNode): List<String> =
        (0 until node.size()).map { index -> node.get(index).requiredText() }

    private fun reviewedSurface(
        case: JsonNode,
        profiles: Map<String, Map<String, JsonNode>>,
    ): ReviewedJavaSourceSurface {
        val reviewedSurfaceProfile =
            profile(
                profiles,
                "reviewed-surface-profiles",
                case.path("reviewedSurfaceProfile").requiredText(),
            )
        val approvalProfile =
            profile(
                profiles,
                "approval-profiles",
                case.path("approvalProfile").requiredText(),
            )
        val budgetVarianceReason =
            if (case.has("budgetVarianceReason")) {
                case.path("budgetVarianceReason").takeUnless(JsonNode::isNull)?.requiredText()
            } else {
                reviewedSurfaceProfile
                    .path("budgetVarianceReason")
                    .takeUnless(JsonNode::isNull)
                    ?.requiredText()
            }
        return reviewedJavaSourceSurface(
            projectPath = FinGrindProjectPaths.CONTRACT,
            relativePath = case.path("relativePath").requiredText(),
            owner = reviewedSurfaceProfile.path("owner").requiredText(),
            reason = "Shared reviewed-surface policy contract case.",
            splitTrigger = reviewedSurfaceProfile.path("splitTrigger").requiredText(),
            roleName = reviewedSurfaceProfile.path("reviewedRoleName").requiredText(),
            budgetVarianceReason = budgetVarianceReason,
            approval = approval(approvalProfile),
        )
    }

    private fun defaultBudget(
        case: JsonNode,
        profiles: Map<String, Map<String, JsonNode>>,
    ): JavaSourceShapeBudget {
        val node =
            profile(
                profiles,
                "budget-profiles",
                case.path("defaultBudgetProfile").requiredText(),
            )
        return JavaSourceShapeBudget(
            roleName = node.path("roleName").requiredText(),
            maxPhysicalLines = node.path("physicalLines").asInt(),
            maxLogicalLines = node.path("logicalLines").asInt(),
            maxImports = node.path("importLikeLines").asInt(),
            maxNestedTypes = node.path("nestedTypes").asInt(),
            maxMethodsPerTopLevelType = node.path("functions").asInt(),
            maxFieldsPerTopLevelType = 1,
            maxSwitchArmsPerMethod = 1,
            maxMethodLineSpan = 10,
            maxMethodParameters = 2,
            maxMethodDecisionPoints = 2,
        )
    }

    private fun approval(node: JsonNode): ReviewedJavaSourceApproval =
        reviewedApproval(
            physicalLines = node.path("physicalLines").asInt(),
            logicalLines = node.path("logicalLines").asInt(),
            imports = node.path("importLikeLines").asInt(),
            nestedTypes = node.path("nestedTypes").asInt(),
            methodsPerTopLevelType = node.path("functions").asInt(),
            fieldsPerTopLevelType = 1,
            switchArmsPerMethod = 1,
            methodLineSpan = 10,
            methodParameters = 2,
            methodDecisionPoints = 2,
            expiresOn = LocalDate.parse(node.path("expiresOn").requiredText()),
        )

    private fun liveMetrics(
        case: JsonNode,
        profiles: Map<String, Map<String, JsonNode>>,
    ): JavaSourceShapeMetrics {
        val node =
            profile(
                profiles,
                "metrics-profiles",
                case.path("liveMetricsProfile").requiredText(),
            )
        return JavaSourceShapeMetrics(
            physicalLineCount = node.path("physicalLines").asInt(),
            logicalLineCount = node.path("logicalLines").asInt(),
            importCount = node.path("importLikeLines").asInt(),
            nestedTypeCount = node.path("nestedTypes").asInt(),
            maxMethodsPerTopLevelType = node.path("functions").asInt(),
            maxFieldsPerTopLevelType = 1,
            maxSwitchArmsPerMethod = 1,
            maxMethodLineSpan = 10,
            maxMethodParameters = 2,
            maxMethodDecisionPoints = 2,
        )
    }

    private fun normalizeViolation(violation: String): String =
        when {
            "without an explicit variance reason" in violation -> "variance-reason-required"
            "is no longer needed because the file fits the" in violation -> "waiver-unnecessary"
            "no longer resolves inside" in violation -> "orphaned-waiver"
            "expired on " in violation ->
                "waiver-expired:${Regex("""expired on (\d{4}-\d{2}-\d{2})""").find(violation)?.groupValues?.get(1)}"
            "no longer matches the live file on " in violation -> normalizeDriftViolation(violation)
            else -> error("Unrecognized reviewed-surface violation: $violation")
        }

    private fun normalizeDriftViolation(violation: String): String {
        val match =
            Regex("""live file on (.+?) \(approved (\d+), live (\d+)\)""")
                .find(violation)
                ?: error("Unrecognized drift violation: $violation")
        val dimension =
            when (match.groupValues[1]) {
                "physical lines" -> "physical-lines"
                "logical lines" -> "logical-lines"
                "imports" -> "imports"
                "nested types" -> "nested-types"
                "methods on one top-level type" -> "functions"
                else -> error("Unsupported shared drift dimension ${match.groupValues[1]}")
            }
        return "snapshot-drift:$dimension:${match.groupValues[2]}:${match.groupValues[3]}"
    }

    private fun JsonNode.requiredText(): String =
        textValue() ?: error("Expected text node in reviewed-surface policy contract: $this")
}

File: scripts/structural_governance/reviewed_surface_policy_contract_test.py
from __future__ import annotations

import json
import re
from datetime import date
from pathlib import Path

from structural_governance.measurement_verification import check_metrics
from structural_governance.models import (
    FileBudget,
    FileMetrics,
    ReviewedSurface,
    ReviewedSurfaceApproval,
)
from structural_governance.reviewed_surface_verification import (
    missing_reviewed_surface_violations,
    reviewed_surface_definition_violations,
    reviewed_surface_violations,
)

_REPO_ROOT = Path(__file__).resolve().parents[2]
_CONTRACT_DIRECTORY = _REPO_ROOT / "scripts/structural_governance/reviewed_surface_policy_contract"

_EXPIRED_PATTERN = re.compile(r"expired on (\d{4}-\d{2}-\d{2})")
_DRIFT_PATTERN = re.compile(r"live file on (.+?) \(approved (\d+), live (\d+)\)")


def _contract_documents() -> list[dict[str, object]]:
    document_paths = sorted(_CONTRACT_DIRECTORY.glob("*.json"))
    if not document_paths:
        raise AssertionError(
            f"No reviewed-surface policy contract cases found in {_CONTRACT_DIRECTORY}"
        )
    return [
        json.loads(document_path.read_text(encoding="utf-8")) for document_path in document_paths
    ]


def _contract_cases() -> list[dict[str, object]]:
    return [
        document for document in _contract_documents() if str(document["documentType"]) == "case"
    ]


def _profiles_by_group() -> dict[str, dict[str, dict[str, object]]]:
    profiles: dict[str, dict[str, dict[str, object]]] = {}
    for document in _contract_documents():
        document_type = str(document["documentType"])
        if document_type == "case":
            continue
        document_profiles = document["profiles"]
        assert isinstance(document_profiles, dict)
        profiles[document_type] = {
            str(profile_name): profile for profile_name, profile in document_profiles.items()
        }
    return profiles


def _profile(
    profiles: dict[str, dict[str, dict[str, object]]],
    profile_group_name: str,
    profile_name: str,
) -> dict[str, object]:
    return profiles[profile_group_name][profile_name]


def _normalize_violation(violation: str) -> str:
    if "without an explicit variance reason" in violation:
        return "variance-reason-required"
    expired_match = _EXPIRED_PATTERN.search(violation)
    if expired_match is not None:
        return f"waiver-expired:{expired_match.group(1)}"
    if "is no longer needed because the file fits the" in violation:
        return "waiver-unnecessary"
    drift_match = _DRIFT_PATTERN.search(violation)
    if drift_match is not None:
        return (
            "snapshot-drift:"
            + _normalize_dimension(drift_match.group(1))
            + f":{drift_match.group(2)}:{drift_match.group(3)}"
        )
    if "no longer resolves inside" in violation:
        return "orphaned-waiver"
    raise AssertionError(f"Unrecognized reviewed-surface violation: {violation}")


def _normalize_dimension(dimension_name: str) -> str:
    return {
        "physical lines": "physical-lines",
        "logical lines": "logical-lines",
        "import-like lines": "imports",
        "functions": "functions",
        "nested types": "nested-types",
    }[dimension_name]


def _normalize_violations(violations: list[str]) -> list[str]:
    return [_normalize_violation(violation) for violation in violations]


def _reviewed_surface(
    case: dict[str, object],
    profiles: dict[str, dict[str, dict[str, object]]],
) -> ReviewedSurface:
    reviewed_surface_profile = _profile(
        profiles,
        "reviewed-surface-profiles",
        str(case["reviewedSurfaceProfile"]),
    )
    approval = _profile(
        profiles,
        "approval-profiles",
        str(case["approvalProfile"]),
    )
    budget_variance_reason = (
        case["budgetVarianceReason"]
        if "budgetVarianceReason" in case
        else reviewed_surface_profile["budgetVarianceReason"]
    )
    assert isinstance(approval, dict)
    return ReviewedSurface(
        relative_path=str(case["relativePath"]),
        owner=str(reviewed_surface_profile["owner"]),
        reason="Shared reviewed-surface policy contract case.",
        split_trigger=str(reviewed_surface_profile["splitTrigger"]),
        reviewed_role_name=str(reviewed_surface_profile["reviewedRoleName"]),
        budget_variance_reason=budget_variance_reason,
        approval=ReviewedSurfaceApproval(
            approved_physical_lines=int(approval["physicalLines"]),
            approved_logical_lines=int(approval["logicalLines"]),
            approved_import_like_lines=int(approval["importLikeLines"]),
            approved_functions=int(approval["functions"]),
            approved_nested_types=int(approval["nestedTypes"]),
            expires_on=date.fromisoformat(str(approval["expiresOn"])),
        ),
    )


def _budget(
    case: dict[str, object],
    profiles: dict[str, dict[str, dict[str, object]]],
) -> FileBudget:
    budget = _profile(
        profiles,
        "budget-profiles",
        str(case["defaultBudgetProfile"]),
    )
    return FileBudget(
        role_name=str(budget["roleName"]),
        max_physical_lines=int(budget["physicalLines"]),
        max_logical_lines=int(budget["logicalLines"]),
        max_import_like_lines=int(budget["importLikeLines"]),
        max_functions=int(budget["functions"]),
        max_nested_types=int(budget["nestedTypes"]),
        max_duplicate_window_lines=30,
        split_hint="Split the example owner.",
    )


def _metrics(
    case: dict[str, object],
    profiles: dict[str, dict[str, dict[str, object]]],
) -> FileMetrics:
    metrics = _profile(
        profiles,
        "metrics-profiles",
        str(case["liveMetricsProfile"]),
    )
    return FileMetrics(
        physical_lines=int(metrics["physicalLines"]),
        logical_lines=int(metrics["logicalLines"]),
        import_like_lines=int(metrics["importLikeLines"]),
        functions=int(metrics["functions"]),
        nested_types=int(metrics["nestedTypes"]),
        normalized_nonempty_lines=tuple("line" for _ in range(int(metrics["logicalLines"]))),
    )


def _run_definition_case(
    case: dict[str, object],
    profiles: dict[str, dict[str, dict[str, object]]],
) -> None:
    reviewed = _reviewed_surface(case, profiles)
    default_budget = _budget(case, profiles)
    actual = _normalize_violations(
        reviewed_surface_definition_violations(
            Path(reviewed.relative_path),
            reviewed,
            default_budget,
        )
    )
    expected = [str(descriptor) for descriptor in case["expectedDescriptors"]]
    if actual != expected:
        raise AssertionError(
            f"Definition case {case['id']} drifted: expected {expected}, actual {actual}"
        )


def _run_runtime_case(
    case: dict[str, object],
    profiles: dict[str, dict[str, dict[str, object]]],
) -> None:
    reviewed = _reviewed_surface(case, profiles)
    default_budget = _budget(case, profiles)
    metrics = _metrics(case, profiles)
    relative_path = Path(reviewed.relative_path)
    baseline_violations = check_metrics(relative_path, default_budget, metrics)
    actual = _normalize_violations(
        reviewed_surface_violations(
            relative_path=relative_path,
            reviewed=reviewed,
            metrics=metrics,
            default_budget=default_budget,
            current_date=date.fromisoformat(str(case["currentDate"])),
            baseline_violations=baseline_violations,
        )
    )
    expected = [str(descriptor) for descriptor in case["expectedDescriptors"]]
    if actual != expected:
        raise AssertionError(
            f"Runtime case {case['id']} drifted: expected {expected}, actual {actual}"
        )


def _run_orphan_case(
    case: dict[str, object],
    profiles: dict[str, dict[str, dict[str, object]]],
) -> None:
    reviewed = _reviewed_surface(case, profiles)
    actual = _normalize_violations(
        missing_reviewed_surface_violations(
            reviewed_surfaces={reviewed.relative_path: reviewed},
            existing_relative_paths={str(path) for path in case["existingRelativePaths"]},
        )
    )
    expected = [str(descriptor) for descriptor in case["expectedDescriptors"]]
    if actual != expected:
        raise AssertionError(
            f"Orphan case {case['id']} drifted: expected {expected}, actual {actual}"
        )


def main() -> int:
    profiles = _profiles_by_group()
    for case in _contract_cases():
        case_type = str(case["caseType"])
        if case_type == "definition":
            _run_definition_case(case, profiles)
            continue
        if case_type == "runtime":
            _run_runtime_case(case, profiles)
            continue
        if case_type == "orphan":
            _run_orphan_case(case, profiles)
            continue
        raise AssertionError(f"Unsupported reviewed-surface policy contract case type: {case_type}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
